// @notionhq/workers/src/capabilities/ai_connector.ts
import { ExecutionError } from "../error.js";
import {
	getPacerState,
	initPacerState,
	type PacerDeclaration,
	type PacerEntry,
} from "../pacer_internal.js";
import type { HandlerOptions, Schedule } from "../types.js";
import type { CapabilityContext } from "./context.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
import {
	parseSchedule,
	type StatefulCapabilityMode,
	type StatefulExecutionResult,
	type RuntimeContext as StatefulRuntimeContext,
} from "./stateful-capability.js";

export type AiConnectorArchetype = "chat" | "project";
/**
 * AI connector syncs are incremental-only for now.
 *
 * Unlike collection syncs, AI connector ingestion does not yet support
 * replace-style reconciliation semantics, so exposing `replace` here would
 * encourage a misleading reset-and-replay model.
 */
export type AiConnectorMode = Exclude<StatefulCapabilityMode, "replace">;

export type AiConnectorChatAuthor = {
	id: string;
	username: string;
};

export type AiConnectorChatMessage = {
	id: string;
	content: string;
	timestamp: string;
	editedTimestamp?: string;
	author: AiConnectorChatAuthor;
};

export type AiConnectorChatChannel = {
	id: string;
	name: string;
};

export type AiConnectorChatRecord = {
	createdAt: string;
	updatedAt: string;
	externalUrl?: string;
	channel: AiConnectorChatChannel;
	threadId?: string;
	messages: AiConnectorChatMessage[];
	chatType?: string;
};

export type AiConnectorProjectBlock = {
	id: string;
	type: string;
	content: string;
};

export type AiConnectorProjectRecord = {
	name: string;
	createdAt: string;
	updatedAt: string;
	externalUrl?: string;
	priority?: string;
	teams?: string[];
	targetEnd?: string;
	createdBy?: {
		id: string;
	};
	blocks: AiConnectorProjectBlock[];
};

export type AiConnectorRecordByArchetype = {
	chat: AiConnectorChatRecord;
	project: AiConnectorProjectRecord;
};

export type AiConnectorChangeUpsert<A extends AiConnectorArchetype> = {
	type: "upsert";
	key: string;
	record: AiConnectorRecordByArchetype[A];
};

export type AiConnectorChange<A extends AiConnectorArchetype> =
	AiConnectorChangeUpsert<A>;

export type AiConnectorExecutionResult<
	A extends AiConnectorArchetype,
	State = unknown,
> = StatefulExecutionResult<AiConnectorChange<A>, State>;

export type AiConnectorConfiguration<
	A extends AiConnectorArchetype,
	State = unknown,
> = {
	/**
	 * The Notion record ID of an existing custom connector in the target workspace.
	 *
	 * This must be the connector record UUID that Notion created, not an external
	 * source identifier.
	 */
	aiConnectorId: string;
	archetype: A;
	mode?: AiConnectorMode;
	schedule?: Schedule;
	execute: (
		state: State | undefined,
		context: CapabilityContext,
	) => Promise<AiConnectorExecutionResult<A, State>>;
};

export type AiConnectorCapability = ReturnType<
	typeof createAiConnectorCapability
>;

type NormalizedAiConnectorConfiguration<A extends AiConnectorArchetype> = {
	aiConnectorId: string;
	archetype: A;
	mode?: AiConnectorMode;
	schedule?: ReturnType<typeof parseSchedule>;
};

type AiConnectorHandlerResult<A extends AiConnectorArchetype, State> = {
	changes: AiConnectorChange<A>[];
	hasMore: boolean;
	nextUserContext?: State;
	nextPacerStates: Record<string, PacerEntry>;
};

export function createAiConnectorCapability<
	A extends AiConnectorArchetype,
	State = unknown,
>(
	key: string,
	configuration: AiConnectorConfiguration<A, State>,
	pacerDeclarations?: readonly PacerDeclaration[],
) {
	return {
		_tag: "ai_connector" as const,
		key,
		config: {
			aiConnectorId: configuration.aiConnectorId,
			archetype: configuration.archetype,
			...(configuration.mode === undefined ? {} : { mode: configuration.mode }),
			...(configuration.schedule === undefined
				? {}
				: { schedule: parseSchedule(configuration.schedule) }),
		} satisfies NormalizedAiConnectorConfiguration<A>,
		async handler(
			runtimeContext?: StatefulRuntimeContext<State>,
			options?: HandlerOptions,
		) {
			const capabilityContext = createCapabilityContext("ai_connector");
			const state = runtimeContext?.state ?? runtimeContext?.userContext;

			initPacerState(runtimeContext?.pacers, pacerDeclarations);

			let executionResult: AiConnectorExecutionResult<A, State>;
			try {
				executionResult = await configuration.execute(state, capabilityContext);
			} catch (err) {
				const error = new ExecutionError(err);
				if (!options?.concreteOutput) {
					writeOutput({
						_tag: "error",
						error: { name: error.name, message: error.message },
					});
				}
				throw error;
			}

			const result = {
				changes: executionResult.changes,
				hasMore: executionResult.hasMore,
				...(executionResult.nextState === undefined
					? {}
					: { nextUserContext: executionResult.nextState }),
				nextPacerStates: getPacerState().pacers,
			} satisfies AiConnectorHandlerResult<A, State>;

			if (options?.concreteOutput) {
				return result;
			}

			writeOutput({ _tag: "success", value: result });

			return result;
		},
	};
}
