// @notionhq/workers/src/capabilities/context.ts
import { Client } from "@notionhq/client";
import type { ClientOptions } from "@notionhq/client/build/src/Client.js";
import type { CapabilityType } from "../worker.js";

export type CapabilityContext = {
	/** Notion API SDK client for this execution. */
	notion: Client;
};

const FIX_STEPS =
	"\n\nTo fix this:\n" +
	"1. Create an internal integration at https://www.notion.so/profile/integrations/internal\n" +
	"2. Give the integration access to the relevant pages/databases\n" +
	"3. Add NOTION_API_TOKEN=<your-token> to your .env file\n" +
	"4. For deployed workers, run: ntn workers env push";

function missingTokenMessage(capabilityType: CapabilityType): string {
	if (capabilityType === "tool") {
		return (
			"NOTION_API_TOKEN is not set. " +
			"When tools are called through a Custom Agent, the platform sets NOTION_API_TOKEN automatically, " +
			"using the permissions of the Custom Agent. " +
			"Outside of that context (e.g. ntn workers exec), you must set it yourself." +
			FIX_STEPS
		);
	}

	return (
		"NOTION_API_TOKEN is not set. " +
		`context.notion requires an API token to make requests in ${capabilityType} capabilities to Notion.` +
		FIX_STEPS
	);
}

function createUnauthenticatedNotionProxy(
	capabilityType: CapabilityType,
): Client {
	return new Proxy({} as Client, {
		get(_target, prop) {
			if (prop === "then" || typeof prop === "symbol") {
				return undefined;
			}
			throw new Error(missingTokenMessage(capabilityType));
		},
	});
}

export function createCapabilityContext(
	capabilityType: CapabilityType,
): CapabilityContext {
	const options: ClientOptions = {};

	if (process.env.NOTION_API_BASE_URL) {
		options.baseUrl = process.env.NOTION_API_BASE_URL;
	}

	if (!process.env.NOTION_API_TOKEN) {
		return { notion: createUnauthenticatedNotionProxy(capabilityType) };
	}

	options.auth = process.env.NOTION_API_TOKEN;
	const notion = new Client(options);

	return { notion };
}
