// @notionhq/workers/src/capabilities/stateful-capability.ts
import { unreachable } from "../error.js";
import type { PacerEntry } from "../pacer_internal.js";
import type { Schedule, SyncSchedule, TimeUnit } from "../types.js";

/**
 * Shared lifecycle mode used by sync-like capabilities.
 */
export type StatefulCapabilityMode = "replace" | "incremental";

/**
 * Base runtime context shared by sync-like capabilities.
 */
export type RuntimeContext<UserContext = unknown> = {
	/** The user-defined/-controlled state (cursor, pagination state, etc.) */
	state?: UserContext;
	/** Legacy field for user-defined/-controlled state. */
	userContext?: UserContext;
	/** Pacer state from the server for rate limiting. */
	pacers?: Record<string, PacerEntry>;
};

/**
 * Shared execution result returned from stateful capability execute functions.
 */
export type StatefulExecutionResult<Change, State = unknown> = {
	changes: Change[];
	hasMore: boolean;
	nextState?: State;
};

const MS_PER_MINUTE = 60 * 1000;
const MS_PER_HOUR = 60 * MS_PER_MINUTE;
const MS_PER_DAY = 24 * MS_PER_HOUR;

const MIN_INTERVAL_MS = MS_PER_MINUTE; // 1m
const MAX_INTERVAL_MS = 7 * MS_PER_DAY; // 7d

/**
 * Parses a user-friendly schedule string into the normalized backend format.
 */
export function parseSchedule(schedule: Schedule): SyncSchedule {
	if (schedule === "continuous") {
		return { type: "continuous" };
	}

	if (schedule === "manual") {
		return { type: "manual" };
	}

	const match = schedule.match(/^(\d+)(m|h|d)$/);
	if (!match || !match[1] || !match[2]) {
		throw new Error(
			`Invalid schedule format: "${schedule}". Use "continuous" or an interval like "30m", "1h", "1d".`,
		);
	}

	const value = parseInt(match[1], 10);
	const unit = match[2] as TimeUnit;

	let intervalMs: number;
	switch (unit) {
		case "m":
			intervalMs = value * MS_PER_MINUTE;
			break;
		case "h":
			intervalMs = value * MS_PER_HOUR;
			break;
		case "d":
			intervalMs = value * MS_PER_DAY;
			break;
		default:
			unreachable(unit);
	}

	if (intervalMs < MIN_INTERVAL_MS) {
		throw new Error(
			`Schedule interval must be at least 1 minute. Got: "${schedule}"`,
		);
	}
	if (intervalMs > MAX_INTERVAL_MS) {
		throw new Error(
			`Schedule interval must be at most 7 days. Got: "${schedule}"`,
		);
	}

	return { type: "interval", intervalMs };
}
