// @notionhq/workers/src/capabilities/sync.ts
import { ExecutionError, RateLimitError } from "../error.js";
import {
	getPacerState,
	initPacerState,
	type PacerDeclaration,
} from "../pacer_internal.js";
import type {
	PropertyConfiguration,
	PropertySchema,
	Schema,
} from "../schema.js";
import type {
	HandlerOptions,
	Icon,
	PeopleValue,
	PlaceValue,
	RelationValue,
	Schedule,
	TextValue,
} from "../types.js";
import type { DatabaseHandle } from "../worker.js";
import type { CapabilityContext } from "./context.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
import {
	parseSchedule,
	type StatefulCapabilityMode,
	type RuntimeContext as StatefulRuntimeContext,
} from "./stateful-capability.js";

type SyncErrorEnvelope =
	| { _tag: "error"; error: { _tag: "generic"; name: string; message: string } }
	| {
			_tag: "error";
			error: {
				_tag: "rate_limit";
				name: string;
				message: string;
				retryAfter?: number;
			};
	  };

/**
 * Maps a property configuration to its corresponding value type.
 */
type PropertyValueType<T extends PropertyConfiguration> = T extends {
	type: "people";
}
	? PeopleValue
	: T extends { type: "place" }
		? PlaceValue
		: T extends { type: "relation" }
			? RelationValue
			: TextValue;

/**
 * Sync mode determines how the sync handles data lifecycle.
 */
export type SyncMode = StatefulCapabilityMode;

/**
 * A change representing a record to be created or updated.
 */
export type SyncChangeUpsert<
	PK extends string,
	S extends PropertySchema<PK>,
> = {
	/**
	 * The type of change. Use `"upsert"` to create or update a record.
	 */
	type: "upsert";
	/**
	 * A unique identifier for this record, used to match against existing pages.
	 */
	key: string;
	/**
	 * The key of the database to write to. If omitted, defaults to
	 * the database associated with the sync capability.
	 */
	targetDatabaseKey?: string;
	/**
	 * The property values for this record.
	 * Keys must match the property names defined in the schema.
	 * Use the Builder helpers (e.g., `Builder.title()`, `Builder.richText()`) to create values.
	 */
	properties: {
		[Property in keyof S]: PropertyValueType<S[Property]>;
	};
	/**
	 * When this record was last updated in the upstream system.
	 * ISO 8601 string (e.g. `"2026-03-19T12:00:00Z"`).
	 * Used for conflict resolution when multiple syncs write to the same database.
	 * Recommended, but optional.
	 */
	upstreamUpdatedAt?: string;
	/**
	 * Optional icon to use as the icon for this row's page.
	 * Use the `Builder.emojiIcon()`, `Builder.notionIcon()`, or `Builder.imageIcon()` helpers.
	 */
	icon?: Icon;
	/**
	 * Optional markdown content to add to the page body.
	 * This will be converted to Notion blocks and added as page content.
	 */
	pageContentMarkdown?: string;
};

/**
 * A change representing a record to be deleted.
 * Only applicable when using `mode: "incremental"`.
 */
export type SyncChangeDelete = {
	/**
	 * The type of change. Use `"delete"` to remove a record.
	 */
	type: "delete";
	/**
	 * The unique identifier of the record to delete.
	 */
	key: string;
	/**
	 * The key of the database to delete from. If omitted, defaults to
	 * the database associated with the sync capability.
	 */
	targetDatabaseKey?: string;
};

/**
 * A change to be applied to the synced database.
 * Can be either an upsert (create/update) or a delete.
 */
export type SyncChange<PK extends string, S extends PropertySchema<PK>> =
	| SyncChangeUpsert<PK, S>
	| SyncChangeDelete;

/**
 * Result returned from the sync execute function.
 */
export type SyncExecutionResult<PK extends string, State = unknown> = {
	/**
	 * The batch of changes to apply in this execution.
	 * Can include upserts (create/update) and deletes.
	 */
	changes: SyncChange<PK, PropertySchema<PK>>[];

	/**
	 * Indicates whether there is more data to fetch.
	 * - `true`: More data available, will trigger another execution with nextState
	 * - `false`: No more data to fetch, sync is complete
	 */
	hasMore: boolean;

	/**
	 * Optional state data to pass to the next execution.
	 * Required if `hasMore` is `true`, ignored if `hasMore` is `false`.
	 * This can be any type of data (cursor, page number, timestamp, etc.).
	 * The same data will be provided as `state` in the next execution.
	 */
	nextState?: State;
};

/**
 * Configuration for a sync capability that syncs data to a database.
 */
export type SyncConfiguration<
	PK extends string,
	S extends Schema<PK>,
	State = unknown,
> = {
	/**
	 * The database to sync data into.
	 */
	database: DatabaseHandle<PK, S>;

	/**
	 * How the sync handles data lifecycle:
	 * - "replace": Each sync returns the complete dataset. After hasMore:false,
	 *              pages not seen in this sync run are deleted.
	 * - "incremental": Sync returns changes only. After hasMore:false, sync
	 *                  continues from saved cursor. Use delete markers
	 *                  to explicitly remove pages.
	 *
	 * @default "replace"
	 */
	mode?: SyncMode;

	/**
	 * How often the sync should run.
	 * - "continuous": Run as frequently as the system allows
	 * - "manual": Only run when explicitly triggered
	 * - Interval string: Run at specified intervals, e.g. "1h", "30m", "1d"
	 *
	 * Minimum interval: 1 minute ("1m")
	 * Maximum interval: 7 days ("7d")
	 *
	 * @default "30m"
	 */
	schedule?: Schedule;

	/**
	 * A function that fetches the data to sync from the third-party service.
	 *
	 * This function can return all data at once, or implement pagination by:
	 * 1. Returning a batch of changes with `hasMore: true` and a `nextState`
	 * 2. The runtime will call execute again with that state
	 * 3. Continue until `hasMore: false` is returned
	 *
	 * The runtime will handle diffing against the data source and creating,
	 * updating, and deleting pages as necessary.
	 *
	 * @param state - User-defined state from the previous execution (undefined on first call)
	 * @param context - Runtime context, including Notion client
	 * @returns A result containing changes, hasMore status, and optional nextState
	 */
	execute: (
		state: State | undefined,
		context: CapabilityContext,
	) => Promise<SyncExecutionResult<PK, State>>;
};

export type SyncCapability = ReturnType<typeof createSyncCapability>;

/**
 * Creates a special handler for syncing third-party data to a database.
 *
 * @param syncConfiguration - The configuration for the sync.
 * @returns A handler function that executes the sync function, and passes data
 * needed to complete the sync back to the platform.
 */
export function createSyncCapability<
	PK extends string,
	S extends Schema<PK>,
	Context = unknown,
>(
	key: string,
	syncConfiguration: SyncConfiguration<PK, S, Context>,
	pacerDeclarations?: readonly PacerDeclaration[],
) {
	return {
		_tag: "sync" as const,
		key,
		config: {
			databaseKey: syncConfiguration.database.key,
			primaryKeyProperty: syncConfiguration.database.config.primaryKeyProperty,
			mode: syncConfiguration.mode,
			schedule: syncConfiguration.schedule
				? parseSchedule(syncConfiguration.schedule)
				: undefined,
		},
		async handler(
			runtimeContext?: StatefulRuntimeContext<Context>,
			options?: HandlerOptions,
		) {
			const capabilityContext = createCapabilityContext("sync");
			const state = runtimeContext?.state ?? runtimeContext?.userContext;

			initPacerState(runtimeContext?.pacers, pacerDeclarations);

			let executionResult: SyncExecutionResult<PK, Context>;
			try {
				executionResult = await syncConfiguration.execute(
					state,
					capabilityContext,
				);
			} catch (err) {
				if (err instanceof RateLimitError) {
					if (!options?.concreteOutput) {
						const envelope: SyncErrorEnvelope = {
							_tag: "error",
							error: {
								_tag: "rate_limit",
								name: err.name,
								message: err.message,
								...(err.retryAfter !== undefined
									? { retryAfter: err.retryAfter }
									: {}),
							},
						};
						writeOutput(envelope);
					}
					throw err;
				}

				const error = new ExecutionError(err);
				if (!options?.concreteOutput) {
					const envelope: SyncErrorEnvelope = {
						_tag: "error",
						error: {
							_tag: "generic",
							name: error.name,
							message: error.message,
						},
					};
					writeOutput(envelope);
				}
				throw error;
			}

			const changes = executionResult.changes.map((change) => ({
				...change,
				targetDatabaseKey:
					change.targetDatabaseKey ?? syncConfiguration.database.key,
			}));

			const result = {
				changes,
				hasMore: executionResult.hasMore,
				nextUserContext: executionResult.nextState,
				nextPacerStates: getPacerState().pacers,
			};

			if (options?.concreteOutput) {
				return result;
			} else {
				writeOutput({ _tag: "success", value: result });
			}

			return result;
		},
	};
}
