// @notionhq/workers/src/capabilities/webhook.ts
import { ExecutionError } from "../error.js";
import type { HandlerOptions } from "../types.js";
import type { CapabilityContext } from "./context.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";

/**
 * A single incoming webhook request.
 */
export interface WebhookEvent {
	/**
	 * Unique ID for this webhook delivery, stable across retries.
	 * Best-effort idempotency key — prefer using the webhook provider's
	 * own delivery/event ID when available, since providers may redeliver
	 * the same event with a new deliveryId.
	 */
	deliveryId: string;
	/**
	 * The parsed JSON body of the incoming request, or an empty object if
	 * the body is not valid JSON.
	 */
	body: Record<string, unknown>;
	/**
	 * The raw request body as a string, useful for signature verification.
	 */
	rawBody: string;
	/**
	 * The HTTP headers from the incoming request
	 */
	headers: Record<string, string>;
	/**
	 * The HTTP method of the incoming request (e.g. "POST")
	 */
	method: string;
}

/**
 * Throw this error from your webhook execute handler to signal that
 * signature verification failed. After 5 consecutive verification
 * failures, the platform will short-circuit and reject all incoming
 * requests for this webhook without executing the handler.
 *
 * @example
 * ```ts
 * worker.webhook("onGithubPush", {
 *   title: "GitHub Push",
 *   description: "Handles GitHub push events",
 *   execute: async (events) => {
 *     const secret = process.env.GITHUB_WEBHOOK_SECRET;
 *     if (!secret) {
 *       throw new WebhookVerificationError("GITHUB_WEBHOOK_SECRET not set");
 *     }
 *     for (const event of events) {
 *       if (!verifyGitHubSignature(event.rawBody, event.headers, secret)) {
 *         throw new WebhookVerificationError("Invalid GitHub signature");
 *       }
 *       // ... handle verified event
 *     }
 *   },
 * });
 * ```
 */
export class WebhookVerificationError extends ExecutionError {
	constructor(cause?: unknown) {
		super(cause ?? "Webhook signature verification failed");
		this.name = "WebhookVerificationError";
	}
}

/**
 * Configuration for a webhook capability
 */
export interface WebhookConfiguration {
	/**
	 * Title of the webhook - shown in the UI when viewing webhooks
	 */
	title: string;

	/**
	 * Description of what this webhook does - shown in the UI
	 */
	description: string;

	/**
	 * The function that executes when the webhook is triggered.
	 * Receives an array of events (currently always a single event,
	 * but may be batched in the future).
	 *
	 * To signal a verification failure, throw a `WebhookVerificationError`.
	 * After 5 consecutive verification failures, the platform will
	 * short-circuit and stop executing the handler.
	 *
	 * @param events - The incoming webhook events
	 * @param context - The capability execution context (Notion client, etc.)
	 * @returns A promise that resolves when the webhook processing completes
	 */
	execute: (
		events: WebhookEvent[],
		context: CapabilityContext,
	) => Promise<void> | void;
}

export type WebhookCapability = ReturnType<typeof createWebhookCapability>;

export function createWebhookCapability(
	key: string,
	config: WebhookConfiguration,
) {
	return {
		_tag: "webhook" as const,
		key,
		config: {
			title: config.title,
			description: config.description,
		},
		async handler(
			events: WebhookEvent[],
			options?: HandlerOptions,
		): Promise<{ status: "success" } | undefined> {
			try {
				const capabilityContext = createCapabilityContext("webhook");
				await config.execute(events, capabilityContext);

				if (options?.concreteOutput) {
					return { status: "success" };
				} else {
					writeOutput({ _tag: "success", value: { status: "success" } });
				}
			} catch (err) {
				const error =
					err instanceof ExecutionError ? err : new ExecutionError(err);

				if (!options?.concreteOutput) {
					writeOutput({
						_tag: "error",
						error: {
							name: error.name,
							message: error.message,
							trace: error.stack,
						},
					});
				}

				throw error;
			}
		},
	};
}
