// @notionhq/workers/src/pacer_internal.ts
/**
 * Internal pacer state helpers used by the runtime.
 * @internal
 */
export type PacerEntry = {
	lastScheduledAtMs: number;
	/** Server-provided (apportioned) request budget. */
	allowedRequests: number;
	/** Server-provided interval window in milliseconds. */
	intervalMs: number;
};

export type PacerState = {
	pacers: Record<string, PacerEntry>;
};

export type PacerDeclaration = {
	readonly key: string;
	readonly config: { allowedRequests: number; intervalMs: number };
};

let pacerState: PacerState = { pacers: {} };

export function setPacerState(state: PacerState): void {
	pacerState = state;
}

export function getPacerState(): PacerState {
	return pacerState;
}

/**
 * Initialize pacer state from runtime context, falling back to the worker's
 * own declarations for local execution.
 * @internal
 */
export function initPacerState(
	runtimePacers: Record<string, PacerEntry> | undefined,
	declarations?: readonly PacerDeclaration[],
): void {
	const pacers = runtimePacers ?? {};
	if (
		Object.keys(pacers).length === 0 &&
		declarations &&
		declarations.length > 0
	) {
		const localPacers: Record<string, PacerEntry> = {};
		for (const decl of declarations) {
			localPacers[decl.key] = {
				lastScheduledAtMs: 0,
				allowedRequests: decl.config.allowedRequests,
				intervalMs: decl.config.intervalMs,
			};
		}
		setPacerState({ pacers: localPacers });
	} else {
		setPacerState({ pacers });
	}
}

/**
 * Core wait logic for pacing API requests.
 * Sleeps if needed to stay under the rate limit.
 * @internal
 */
export async function pacerWait(key: string): Promise<void> {
	const state = getPacerState();
	const entry = state.pacers[key];

	if (!entry) {
		throw new Error(
			`Pacer "${key}" not found. Make sure you declared it with worker.pacer("${key}", ...) ` +
				`and are running inside the worker runtime.`,
		);
	}

	const { allowedRequests, intervalMs } = entry;
	const paceMs = Math.ceil(intervalMs / allowedRequests);
	const now = Date.now();
	const lastScheduledAtMs = entry.lastScheduledAtMs;

	// Schedule at the later of: (last scheduled + pace) or now
	const scheduledAtMs = Math.max(lastScheduledAtMs + paceMs, now);
	const delayMs = scheduledAtMs - now;

	// Update state with new scheduled timestamp
	state.pacers[key] = { ...entry, lastScheduledAtMs: scheduledAtMs };
	setPacerState(state);

	// Sleep if needed
	if (delayMs > 0) {
		await new Promise((resolve) => setTimeout(resolve, delayMs));
	}
}
