// @notionhq/workers/dist/capabilities/ai_connector.js
import { ExecutionError } from "../error.js";
import {
  getPacerState,
  initPacerState
} from "../pacer_internal.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
import {
  parseSchedule
} from "./stateful-capability.js";
function createAiConnectorCapability(key, configuration, pacerDeclarations) {
  return {
    _tag: "ai_connector",
    key,
    config: {
      aiConnectorId: configuration.aiConnectorId,
      archetype: configuration.archetype,
      ...configuration.mode === void 0 ? {} : { mode: configuration.mode },
      ...configuration.schedule === void 0 ? {} : { schedule: parseSchedule(configuration.schedule) }
    },
    async handler(runtimeContext, options) {
      const capabilityContext = createCapabilityContext("ai_connector");
      const state = runtimeContext?.state ?? runtimeContext?.userContext;
      initPacerState(runtimeContext?.pacers, pacerDeclarations);
      let executionResult;
      try {
        executionResult = await configuration.execute(state, capabilityContext);
      } catch (err) {
        const error = new ExecutionError(err);
        if (!options?.concreteOutput) {
          writeOutput({
            _tag: "error",
            error: { name: error.name, message: error.message }
          });
        }
        throw error;
      }
      const result = {
        changes: executionResult.changes,
        hasMore: executionResult.hasMore,
        ...executionResult.nextState === void 0 ? {} : { nextUserContext: executionResult.nextState },
        nextPacerStates: getPacerState().pacers
      };
      if (options?.concreteOutput) {
        return result;
      }
      writeOutput({ _tag: "success", value: result });
      return result;
    }
  };
}
export {
  createAiConnectorCapability
};
