// @notionhq/workers/dist/capabilities/automation.js
import { ExecutionError } from "../error.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
function createAutomationCapability(key, config) {
  return {
    _tag: "automation",
    key,
    config: {
      title: config.title,
      description: config.description
    },
    async handler(event, options) {
      try {
        const capabilityContext = createCapabilityContext("automation");
        await config.execute(event, capabilityContext);
        if (options?.concreteOutput) {
          return { status: "success" };
        } else {
          writeOutput({ _tag: "success", value: { status: "success" } });
        }
      } catch (err) {
        const error = new ExecutionError(err);
        if (!options?.concreteOutput) {
          writeOutput({
            _tag: "error",
            error: {
              name: error.name,
              message: error.message,
              trace: error.stack
            }
          });
        }
        throw error;
      }
    }
  };
}
export {
  createAutomationCapability
};
