// @notionhq/workers/dist/capabilities/sync.js
import { ExecutionError, RateLimitError } from "../error.js";
import {
  getPacerState,
  initPacerState
} from "../pacer_internal.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
import {
  parseSchedule
} from "./stateful-capability.js";
function createSyncCapability(key, syncConfiguration, pacerDeclarations) {
  return {
    _tag: "sync",
    key,
    config: {
      databaseKey: syncConfiguration.database.key,
      primaryKeyProperty: syncConfiguration.database.config.primaryKeyProperty,
      mode: syncConfiguration.mode,
      schedule: syncConfiguration.schedule ? parseSchedule(syncConfiguration.schedule) : void 0
    },
    async handler(runtimeContext, options) {
      const capabilityContext = createCapabilityContext("sync");
      const state = runtimeContext?.state ?? runtimeContext?.userContext;
      initPacerState(runtimeContext?.pacers, pacerDeclarations);
      let executionResult;
      try {
        executionResult = await syncConfiguration.execute(
          state,
          capabilityContext
        );
      } catch (err) {
        if (err instanceof RateLimitError) {
          if (!options?.concreteOutput) {
            const envelope = {
              _tag: "error",
              error: {
                _tag: "rate_limit",
                name: err.name,
                message: err.message,
                ...err.retryAfter !== void 0 ? { retryAfter: err.retryAfter } : {}
              }
            };
            writeOutput(envelope);
          }
          throw err;
        }
        const error = new ExecutionError(err);
        if (!options?.concreteOutput) {
          const envelope = {
            _tag: "error",
            error: {
              _tag: "generic",
              name: error.name,
              message: error.message
            }
          };
          writeOutput(envelope);
        }
        throw error;
      }
      const changes = executionResult.changes.map((change) => ({
        ...change,
        targetDatabaseKey: change.targetDatabaseKey ?? syncConfiguration.database.key
      }));
      const result = {
        changes,
        hasMore: executionResult.hasMore,
        nextUserContext: executionResult.nextState,
        nextPacerStates: getPacerState().pacers
      };
      if (options?.concreteOutput) {
        return result;
      } else {
        writeOutput({ _tag: "success", value: result });
      }
      return result;
    }
  };
}
export {
  createSyncCapability
};
