// @notionhq/workers/dist/capabilities/webhook.js
import { ExecutionError } from "../error.js";
import { createCapabilityContext } from "./context.js";
import { writeOutput } from "./output.js";
class WebhookVerificationError extends ExecutionError {
  constructor(cause) {
    super(cause ?? "Webhook signature verification failed");
    this.name = "WebhookVerificationError";
  }
}
function createWebhookCapability(key, config) {
  return {
    _tag: "webhook",
    key,
    config: {
      title: config.title,
      description: config.description
    },
    async handler(events, options) {
      try {
        const capabilityContext = createCapabilityContext("webhook");
        await config.execute(events, capabilityContext);
        if (options?.concreteOutput) {
          return { status: "success" };
        } else {
          writeOutput({ _tag: "success", value: { status: "success" } });
        }
      } catch (err) {
        const error = err instanceof ExecutionError ? err : new ExecutionError(err);
        if (!options?.concreteOutput) {
          writeOutput({
            _tag: "error",
            error: {
              name: error.name,
              message: error.message,
              trace: error.stack
            }
          });
        }
        throw error;
      }
    }
  };
}
export {
  WebhookVerificationError,
  createWebhookCapability
};
