// @notionhq/workers/dist/pacer_internal.js
let pacerState = { pacers: {} };
function setPacerState(state) {
  pacerState = state;
}
function getPacerState() {
  return pacerState;
}
function initPacerState(runtimePacers, declarations) {
  const pacers = runtimePacers ?? {};
  if (Object.keys(pacers).length === 0 && declarations && declarations.length > 0) {
    const localPacers = {};
    for (const decl of declarations) {
      localPacers[decl.key] = {
        lastScheduledAtMs: 0,
        allowedRequests: decl.config.allowedRequests,
        intervalMs: decl.config.intervalMs
      };
    }
    setPacerState({ pacers: localPacers });
  } else {
    setPacerState({ pacers });
  }
}
async function pacerWait(key) {
  const state = getPacerState();
  const entry = state.pacers[key];
  if (!entry) {
    throw new Error(
      `Pacer "${key}" not found. Make sure you declared it with worker.pacer("${key}", ...) and are running inside the worker runtime.`
    );
  }
  const { allowedRequests, intervalMs } = entry;
  const paceMs = Math.ceil(intervalMs / allowedRequests);
  const now = Date.now();
  const lastScheduledAtMs = entry.lastScheduledAtMs;
  const scheduledAtMs = Math.max(lastScheduledAtMs + paceMs, now);
  const delayMs = scheduledAtMs - now;
  state.pacers[key] = { ...entry, lastScheduledAtMs: scheduledAtMs };
  setPacerState(state);
  if (delayMs > 0) {
    await new Promise((resolve) => setTimeout(resolve, delayMs));
  }
}
export {
  getPacerState,
  initPacerState,
  pacerWait,
  setPacerState
};
