// @notionhq/workers/dist/worker.js
import { createAiConnectorCapability } from "./capabilities/ai_connector.js";
import { createAutomationCapability } from "./capabilities/automation.js";
import { createOAuthCapability } from "./capabilities/oauth.js";
import { createSyncCapability } from "./capabilities/sync.js";
import { createToolCapability } from "./capabilities/tool.js";
import {
  createWebhookCapability,
  WebhookVerificationError
} from "./capabilities/webhook.js";
import { pacerWait } from "./pacer_internal.js";
import { createRequire } from "node:module";
const SDK_VERSION = (() => {
  const require2 = createRequire(import.meta.url);
  const packageJson = require2("../package.json");
  if (typeof packageJson !== "object" || packageJson === null || !("version" in packageJson) || typeof packageJson.version !== "string") {
    throw new Error("Failed to read SDK version from package.json");
  }
  return packageJson.version;
})();
class Worker {
  #capabilities = /* @__PURE__ */ new Map();
  #databases = /* @__PURE__ */ new Map();
  #pacers = /* @__PURE__ */ new Map();
  database(key, config) {
    this.#validateUniqueKey(key);
    const database = { key, config };
    this.#databases.set(key, { key, config });
    return database;
  }
  /**
   * Register a pacer for rate-limiting requests to an external API.
   *
   * Multiple syncs can share a pacer — the server will apportion the
   * budget evenly across all syncs that reference it.
   *
   * Example:
   *
   * ```ts
   * const jiraPacer = worker.pacer("jira", {
   *   allowedRequests: 10,
   *   intervalMs: 1000,
   * });
   *
   * worker.sync("jiraSync", {
   *   database: tasks,
   *   execute: async (state, { notion }) => {
   *     await jiraPacer.wait();
   *     const data = await fetchFromJira();
   *     // ...
   *   },
   * });
   * ```
   *
   * @param key - The unique key for this pacer.
   * @param config - The rate limit configuration.
   * @returns A pacer handle. Call `handle.wait()` before each API request.
   */
  pacer(key, config) {
    this.#validateUniqueKey(key);
    this.#pacers.set(key, { key, config });
    return {
      key,
      config,
      wait: () => pacerWait(key)
    };
  }
  /**
   * Register a sync capability.
   *
   * Example:
   *
   * ```ts
   * import { Worker } from "@notionhq/workers";
   * import * as Builder from "@notionhq/workers/builder";
   * import * as Schema from "@notionhq/workers/schema";
   *
   * const worker = new Worker();
   * export default worker;
   *
   * const tasks = worker.database("tasks", {
   *   type: "managed",
   *   initialTitle: "Tasks",
   *   schema: {
   *     properties: {
   *       "Task Name": Schema.title(),
   *       Status: Schema.select([
   *         { name: "Open", color: "default" },
   *         { name: "Done", color: "green" },
   *       ]),
   *     },
   *   },
   * });
   *
   * worker.sync("tasksSync", {
   *   database: tasks,
   *   execute: async () => {
   *     const changes = [
   *       {
   *         key: "task-1",
   *         type: "upsert",
   *         properties: {
   *           "Task Name": Builder.title("Write docs"),
   *           Status: Builder.select("Open"),
   *         },
   *       },
   *     ];
   *
   *     return { changes, hasMore: false };
   *   },
   * });
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The sync configuration.
   * @returns The capability object.
   */
  sync(key, config) {
    this.#validateUniqueKey(key);
    const capability = createSyncCapability(
      key,
      config,
      Array.from(this.#pacers.values())
    );
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register an AI connector capability.
   *
   * Example:
   *
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.aiConnector("supportChats", {
   *   aiConnectorId: "connector-record-id",
   *   archetype: "chat",
   *   mode: "incremental",
   *   schedule: "15m",
   *   execute: async (state) => ({
   *     changes: [
   *       {
   *         type: "upsert",
   *         key: "thread-1",
   *         record: {
   *           createdAt: "2026-01-01T00:00:00Z",
   *           updatedAt: "2026-01-01T00:00:00Z",
   *           channel: {
   *             id: "channel-1",
   *             name: "Support",
   *           },
   *           threadId: "thread-1",
   *           messages: [
   *             {
   *               id: "message-1",
   *               content: "Hello from support",
   *               timestamp: "2026-01-01T00:00:00Z",
   *               author: {
   *                 id: "user-1",
   *                 username: "alice",
   *               },
   *             },
   *           ],
   *         },
   *       },
   *     ],
   *     hasMore: false,
   *   }),
   * });
   * ```
   *
   * `aiConnectorId` must be the Notion record ID of an existing custom
   * connector in the workspace where you deploy this worker.
   *
   * @param key - The unique key for this capability.
   * @param config - The AI connector configuration.
   * @returns The capability object.
   */
  aiConnector(key, config) {
    this.#validateUniqueKey(key);
    const capability = createAiConnectorCapability(
      key,
      config,
      Array.from(this.#pacers.values())
    );
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register a tool capability.
   *
   * Example:
   *
   * ```ts
   * import { j } from "@notionhq/workers/schema-builder";
   *
   * worker.tool<{ name: string }, string>("sayHello", {
   *   title: "Say Hello",
   *   description: "Say hello to the user",
   *   schema: j.object({ name: j.string() }),
   *   execute: ({ name }, { notion }) => {
   *     return `Hello, ${name}!`;
   *   },
   * })
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The tool configuration.
   * @returns The capability object.
   */
  tool(key, config) {
    this.#validateUniqueKey(key);
    const capability = createToolCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register an automation capability.
   *
   * Example:
   *
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.automation("sendWelcomeEmail", {
   *   title: "Send Welcome Email",
   *   description: "Sends a welcome email when a new user is added",
   *   execute: async (event, { notion }) => {
   *     const { pageId, pageData } = event;
   *
   *     // Access page properties from the Public API format
   *     if (pageData) {
   *       const name = pageData.properties.Name; // Access any property
   *       const status = pageData.properties.Status;
   *       console.log(`Processing: ${name}`);
   *     }
   *
   *     // Your automation logic here
   *     await sendEmail(pageId);
   *   },
   * })
   * ```
   *
   * @param key - The unique key for this capability.
   * @param config - The automation configuration.
   * @returns The capability object.
   */
  automation(key, config) {
    this.#validateUniqueKey(key);
    const capability = createAutomationCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register a webhook capability.
   *
   * Webhooks expose an HTTP endpoint that external services can call.
   * When hit, the request is processed by the worker's execute function.
   *
   * Example:
   *
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.webhook("onGithubPush", {
   *   title: "GitHub Push Webhook",
   *   description: "Handles push events from GitHub",
   *   execute: async (events, { notion }) => {
   *    for (const event of events) {
   *      console.log("Received webhook:", event.body);
   *      console.log("Headers:", event.headers);
   *    }
   *   },
   * })
   * ```
   *
   * @param key - The unique key for this capability. Used as the webhook name in the URL.
   * @param config - The webhook configuration.
   * @returns The capability object.
   */
  webhook(key, config) {
    this.#validateUniqueKey(key);
    const capability = createWebhookCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Register an OAuth capability.
   *
   * There are two ways to configure OAuth:
   *
   * 1. Notion-managed providers:
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.oauth("googleAuth", {
   *   type: "notion_managed",
   *   name: "my-google-auth",
   *   provider: "google"
   * })
   * ```
   *
   * 2. User-managed OAuth configuration:
   * ```ts
   * const worker = new Worker();
   * export default worker;
   *
   * worker.oauth("myCustomAuth", {
   *   type: "user_managed",
   *   name: "my-custom-oauth",
   *   authorizationEndpoint: "https://provider.com/oauth/authorize",
   *   tokenEndpoint: "https://provider.com/oauth/token",
   *   scope: "read write",
   *   clientId: process.env.CLIENT_ID,
   *   clientSecret: process.env.CLIENT_SECRET,
   *   authorizationParams: {
   *     access_type: "offline",
   *     prompt: "consent"
   *   }
   * })
   * ```
   *
   * @param key - The unique key used to register this OAuth capability.
   * @param config - The OAuth configuration (Notion-managed or user-managed) for this capability.
   * @returns The registered OAuth capability.
   */
  oauth(key, config) {
    this.#validateUniqueKey(key);
    const capability = createOAuthCapability(key, config);
    this.#capabilities.set(key, capability);
    return capability;
  }
  /**
   * Get all registered capabilities (for discovery) without their handlers.
   */
  get capabilities() {
    return Array.from(this.#capabilities.values()).map((c) => ({
      _tag: c._tag,
      key: c.key,
      config: c.config
    }));
  }
  get manifest() {
    return {
      sdkVersion: SDK_VERSION,
      databases: Array.from(this.#databases.values()),
      pacers: Array.from(this.#pacers.values()),
      capabilities: this.capabilities
    };
  }
  /**
   * Execute a capability by key.
   *
   * @param key - The key of the capability to execute.
   * @param context - The context to pass to the capability.
   * @param options - Additional options for execution (e.g. for testing).
   * @returns The result of the capability execution.
   */
  async run(key, context, options = {}) {
    const capability = this.#capabilities.get(key);
    if (!capability) {
      throw new Error(`Capability "${key}" not found`);
    }
    if (capability._tag === "oauth") {
      throw new Error(
        `Cannot run OAuth capability "${key}" - OAuth capabilities only provide configuration`
      );
    }
    if (!capability.handler) {
      throw new Error(`Capability "${key}" cannot be executed`);
    }
    return capability.handler(context, options);
  }
  #validateUniqueKey(key) {
    if (!key || typeof key !== "string") {
      throw new Error("Capability key must be a non-empty string");
    }
    if (this.#capabilities.has(key) || this.#databases.has(key) || this.#pacers.has(key)) {
      throw new Error(`Worker item with key "${key}" already registered`);
    }
  }
}
export {
  WebhookVerificationError,
  Worker
};
